### PR TITLE
docs: Add FAQ on reasons for need to downgrade dependencies

### LINKED ISSUE #1513: docs: Mention that downgrading dependencies might be needed at times
# Description

Given the problems brought up with `click` `v8.X` in Issue #1506, it is probably worth noting in the docs FAQ section that if a user runs into a problem with a core dependency acting weird that as `pyhf` is a library and not an application we don't overly control the range of dependencies and they might need to manually downgrade dependencies, e.g.

```
# requirements.txt
click<8.0.0
pyhf==0.6.2
```

, especially if there has been a recent release of the dependencies since the last release of `pyhf`.

### PR DESCRIPTION
**Pull request by matthewfeickert**
# Description

Resolves #1513

Add FAQ explaining how reasons why users might have to manually downgrade dependencies.

# Checklist Before Requesting Reviewer

- [x] Tests are passing
- [x] "WIP" removed from the title of the pull request
- [x] Selected an Assignee for the PR to be responsible for the log summary

# Before Merging

For the PR Assignees:

- [x] Summarize commit messages into a comprehensive review of the PR

```
* Add FAQ explaining how reasons why users might have to manually downgrade dependencies.
   - c.f. PR #1979 for context
```

### CODE REVIEW

**Comment by matthewfeickert** [comment]
Relevant RTD page: https://pyhf--1529.org.readthedocs.build/en/1529/faq.html#i-installed-pyhf-from-pypi-why-am-i-getting-an-error-from-a-dependency

**Comment by matthewfeickert** [comment]
@alexander-held @henryiii if you also have time to let me know if this seems clear (and not too verbose) that would nice.

**Review by kratsg — APPROVED** [review]

File: docs/faq.rst [hunk 1]
```diff
@@ -28,6 +28,39 @@ Use the :code:`--backend` option for :code:`pyhf cls` to specify a tensor backen
...
+``pyhf`` and other libraries defined in an environment file (i.e.,
+``requirements.txt``) could have directly conflicting dependencies that would
+result in ``pip`` failing to be able to install ``pyhf``.
+
+To give an explicit example, changes in ``click``'s behavior between its
+``v7.X`` releases and ``v8.X`` resulted in a runtime ``TypeError`` if ``click``
+``v8.X`` was used with ``pyhf`` ``v0.6.2`` (c.f. Issue :issue:`1506`).
+The Issue was fixed in the next release of ``pyhf``, but the intermediate
+solution for users was to simply install an older version of ``click`` that was
+still compatible with ``pyhf``
```
Comment (alexander-held) [inline docs/faq.rst:56]: This is missing a period/colon after `pyhf`.

**Comment by alexander-held** [comment]
This looks good to me and seems helpful to have. What do you think about adding the required `pip` command for the example? Something like
```bash
python -m pip install --upgrade "click<8"
```
for the `click` example. Maybe being very explicit here could help users who are getting started with Python and have not had to deal with specifying versions before. On the other hand, this is not a Python tutorial so it also seems reasonable to not go into that detail. The instructions can be found elsewhere on the internet.

**Comment by henryiii** [comment]
To me this looks more like a bug in Click that happened to get added in version 8 than an intentional change. Once it's discovered I think you might want to do a bug report to Click. I'm not sure a document like this is all that useful; this is mostly not specific to PyHF, or any other library that has dependencies. If you depend on something, that something could change - or even just not work on a new platform. Even one that "caps" all dependencies is susceptible to bugs creeping into dependencies, etc. If you provide a "known working version lockfile" perhaps, but in general, as long as the library is being actively developed, this shouldn't be a problem.

**Comment by matthewfeickert** [comment]
> Maybe being very explicit here could help users who are getting started with Python and have not had to deal with specifying versions before.

@alexander-held This is somewhat the reason that I put them in a `requirements.txt` as an example to indicate that there was a relationship between them. I'd probably prefer to not give too many commands here though.

> To me this looks more like a bug in Click that happened to get added in version 8 than an intentional change. Once it's discovered I think you might want to do a bug report to Click.

@henryiii if you have time can you elaborate a bit more on this (or maybe over on Issue #1506)? There was some [past discussion on the IRIS-HEP Slack](https://iris-hep.slack.com/archives/CDGH6HKMG/p1625068565357600) where it seemed like this was our fault due to using a weakref.

> If you depend on something, that something could change - or even just not work on a new platform. Even one that "caps" all dependencies is susceptible to bugs creeping into dependencies, etc.

Here you're talking about users needing to be responsible for controlling their own dependencies for an application, correct? If so, that is what I was trying to convey.

> If you provide a "known working version lockfile" perhaps, but in general, as long as the library is being actively developed, this shouldn't be a problem.

I agree that most of the time it shouldn't be an issue, but we're running into this now and while we've been busy with conferences the last few weeks `pyhf` is decently actively developed and Alex's original issue is well over two weeks old at this point (though this is also the first time we've run into something like this).

**Comment by matthewfeickert** [comment]
Relevant RTD page: https://pyhf--1529.org.readthedocs.build/en/1529/faq.html#i-installed-an-old-pyhf-release-from-pypi-why-am-i-getting-an-error-from-a-dependency

**Pull request merged**